src/main/scala/org/bitcoins/core/util/BitcoinScriptUtil.scala

```diff
@@ -372,12 +343,41 @@ trait BitcoinScriptUtil extends BitcoinSLogger {
     } else program.originalScript
   }
 
-  def parseScriptEither(scriptEither: Either[(Seq[ScriptToken], ScriptPubKey), ScriptError]): Seq[ScriptToken] = scriptEither match {
+  private def parseScriptEither(scriptEither: Either[(Seq[ScriptToken], ScriptPubKey), ScriptError]): Seq[ScriptToken] = scriptEither match {
     case Left((_,scriptPubKey)) =>
       logger.debug("Script pubkey asm inside calculateForSigning: " + scriptPubKey.asm)
       scriptPubKey.asm
     case Right(_) => Nil //error
   }
+
+  /** Given a tx, scriptPubKey and the input index we are checking the tx, it derives the appropriate [[SignatureVersion]] to use */
+  def parseSigVersion(tx: Transaction, scriptPubKey: ScriptPubKey, inputIndex: UInt32): SignatureVersion  = scriptPubKey match {
+    case _ : WitnessScriptPubKeyV0 | _: UnassignedWitnessScriptPubKey =>
+      SigVersionWitnessV0
+    case _: P2SHScriptPubKey =>
+      val scriptSig = tx.inputs(inputIndex.toInt).scriptSignature
+      scriptSig match {
+        case s : P2SHScriptSignature =>
+          parseSigVersion(tx,s.redeemScript,inputIndex)
+        case _ : P2PKScriptSignature | _: P2PKHScriptSignature | _:MultiSignatureScriptSignature | _: NonStandardScriptSignature
+          | _: CLTVScriptSignature | _: CSVScriptSignature | EmptyScriptSignature => SigVersionBase
+      }
+    case _: P2PKScriptPubKey | _: P2PKHScriptPubKey | _: MultiSignatureScriptPubKey  | _: NonStandardScriptPubKey
+         | _: CLTVScriptPubKey | _: CSVScriptPubKey | EmptyScriptPubKey => SigVersionBase
+  }
+
+
+  /** Casts the given script token to a boolean value
+    * Mimics this function inside of Bitcoin Core
+    * [[https://github.com/bitcoin/bitcoin/blob/8c1dbc5e9ddbafb77e60e8c4e6eb275a3a76ac12/src/script/interpreter.cpp#L38]]
+    * All bytes in the byte vector must be zero, unless it is the last byte, which can be 0 or 0x80 (negative zero)
+    * */
+  def castToBool(token: ScriptToken): Boolean = {
+    token.bytes.zipWithIndex.exists {
+      case (b,index) =>
+        b.toByte != 0 && !(b.toByte == 0x80.toByte && index == token.bytes.size - 1)
+    }
```

Comment:
Maybe just me, but I think this is cleaner:

    val byteNotZero = b.toByte != 0
    val lastByteNotNegativeZero = !(b.toByte == 0x80.toByte && index == token.bytes.size - 1)
    byteNotZero && lastByteNotNegativeZero